Get the name of the joints from the model

buschbapti opened the pull request:
The name of the joints in the `Jacobian` were set to their default value `{"joint0", "joint1", ..., "jointN"}`. We can access them from the model and this PR is setting the correct joint names to the `Jacobian` at compute time.

domire8 reviewed (approved):
To be honest, I still don't like it but if you need this, then I can live with it. 
We are excluding the universe joint from the list of joint names and the number of joints, but nowhere else (for example frames).

buschbapti commented:
I think the problem is actually broader than that. They include fixed joints in joint name I guess and we don't want them either. At least not in the jacobian. We have to look if there is a get_active_joints or something equivalent

eeberhard commented:
We might want the Jacobian for some arbitrary end-effector, which might not be defined by a _link_ joint but rather by a _frame_ (including e.g. fixed joints). I don't think that feature is in the scope of this PR, but I do see it as a real user need in the future. Do you think this PR goes towards or against that higher level idea?

eeberhard commented:
Also slightly out of scope comment: StateRep::Jacobian stores a list of joint names, which I see is used in the constructor as a way of counting the joint numbers and thereby matrix size, and perhaps it's nice to attribute the elements of the joint-space dimension of the matrix to their names. But it doesn't store the frame name in the task-space dimension of the matrix - so it seems a bit unbalanced to me. It doesn't need a list of joint names if it's just a wrapper for a matrix, but if it is storing joint-space frame names then it should also store the task-space frame name. Mainly, it will not always be the case that the last joint is also the frame that is being mapped by the Jacobian.

buschbapti commented:
Yes I totally agree with you, `Jacobian` should also includes those info. It is a bit outside of this PR though as we need to refactor the `Jacobian` class. 

I think we need a better high level goal for the `joint_names` idea. Basically, it is used not only to check the size of the matrix but also that the multiplication with a `JointState` make sense. The main idea was that, sometimes, you get a `JointState` where the names are in a different order than the one in the model (hello Universal Robot). So with this, when multiplying with the `Jacobian` it will fail instead of giving a result that does not make sense.

buschbapti commented:
> We might want the Jacobian for some arbitrary end-effector, which might not be defined by a _link_ joint but rather by a _frame_ (including e.g. fixed joints). I don't think that feature is in the scope of this PR, but I do see it as a real user need in the future. Do you think this PR goes towards or against that higher level idea?

Well the problem is that Pinocchio is actually not storing fixed joints (except for this universe joint). So we can't even do that with the lib at the moment.

eeberhard commented:
> > We might want the Jacobian for some arbitrary end-effector, which might not be defined by a _link_ joint but rather by a _frame_ (including e.g. fixed joints). I don't think that feature is in the scope of this PR, but I do see it as a real user need in the future. Do you think this PR goes towards or against that higher level idea?
> 
> Well the problem is that Pinocchio is actually not storing fixed joints (except for this universe joint). So we can't even do that with the lib at the moment.

Well, I do think it's fair to store joint names and `nq` for the joint-space side of the Jacobian. And I don't want to bring state representation issues / features into the scope of this PR. But, I'm starting to think, if you're adding a `get_joint_names` method to joint names, then I think it's related and useful to add a `get_frame_names` method. We have access to `model.frames[idx].name` after all. This makes the Model less "joint-centric" and supports the extension of the Jacobian class in future, and make the whole "joint" vs "frames" issue more explicit.

eeberhard commented on source/robot_model/include/robot_model/Model.hpp:
```diff
@@ -208,7 +214,14 @@ inline void Model::set_urdf_path(const std::string& urdf_path) {
 
 inline unsigned int Model::get_nb_joints() const {
```
btw is `nb` short for "number"? That's not a common abbreviation in my opinion. Maybe it could be refactored to `get_number_of_joints` or the very verbose `get_joint_degrees_of_freedom` or the very sparse `get_nq`. Just some thoughts, I don't have a strong preference but I don't think `nb_joints` is very clear on its own.

buschbapti replied:
yes make sense. I think the verbose `get_number_of_joints` is the better choice here

buschbapti commented:
> Well, I do think it's fair to store joint names and `nq` for the joint-space side of the Jacobian. And I don't want to bring state representation issues / features into the scope of this PR. But, I'm starting to think, if you're adding a `get_joint_names` method to joint names, then I think it's related and useful to add a `get_frame_names` method. We have access to `model.frames[idx].name` after all. This makes the Model less "joint-centric" and supports the extension of the Jacobian class in future, and make the whole "joint" vs "frames" issue more explicit.

Oh I see yes that make totally sense at this should be in the scope of this PR. I will add that

domire8 commented:
> > > We might want the Jacobian for some arbitrary end-effector, which might not be defined by a _link_ joint but rather by a _frame_ (including e.g. fixed joints). I don't think that feature is in the scope of this PR, but I do see it as a real user need in the future. Do you think this PR goes towards or against that higher level idea?
> > 
> > 
> > Well the problem is that Pinocchio is actually not storing fixed joints (except for this universe joint). So we can't even do that with the lib at the moment.
> 
> Well, I do think it's fair to store joint names and `nq` for the joint-space side of the Jacobian. And I don't want to bring state representation issues / features into the scope of this PR. But, I'm starting to think, if you're adding a `get_joint_names` method to joint names, then I think it's related and useful to add a `get_frame_names` method. We have access to `model.frames[idx].name` after all. This makes the Model less "joint-centric" and supports the extension of the Jacobian class in future, and make the whole "joint" vs "frames" issue more explicit.

Yes I totally agree with this. A `get_frame_names` will be absolutely necessary at some point. We have to treat both equivalently and consistently. I think due to the fact that the lib is not used a lot so far, we have troubles defining our requirements.

buschbapti commented:
Alright so I have added the `get_frames` function and rename the `get_joint_names` to `get_joint_frames` for consistency with the functions. As discussed I have discarded the `universe` and `root_joint` frames. I have also moved the function that where taking `frame_ids` as private as a user has no way of knowing the id associated to a frame anyway.

eeberhard commented on source/robot_model/src/Model.cpp:
```diff
@@ -41,6 +41,17 @@ Model::Model(const Model& model) :
+  for (auto& f : this->robot_model_.frames) {
+    frames.push_back(f.name);
+  }
+  this->frame_names_ = std::vector<std::string>(frames.begin() + 2, frames.end());
```
maybe add a brief comment to say that you are intentionally skipping `universe` and `root_joint`? (since in get_joint_frames you justify omitting universe)

buschbapti replied:
indeed thanks

eeberhard commented on source/robot_model/include/robot_model/Model.hpp:
```diff
@@ -206,14 +219,20 @@ inline void Model::set_urdf_path(const std::string& urdf_path) {
+}
+
+inline std::vector<std::string> Model::get_joint_frames() const {
+  // model contains a first joint called universe that needs to be discarded
```
Is the first joint called `universe` or `root_joint`? not that it matters too much for this comment. I appreciate that you are showing that the +1 is intentional

domire8 replied:
Its called `universe`

eeberhard reviewed (approved):
Nice work, and thanks for opening the discussion and taking the feedback on board. Seems good to me!

I noticed you added the new urdf model as a test fixture, but it's not being used by anything. Maybe you can unstage it until a test actually needs this. Then again it's not massive and I suspect we will indeed want it soon.

domire8 commented on source/robot_model/include/robot_model/Model.hpp:
```diff
@@ -206,14 +219,20 @@ inline void Model::set_urdf_path(const std::string& urdf_path) {
 
-inline unsigned int Model::get_nb_joints() const {
+inline unsigned int Model::get_number_of_joints() const {
   // subtract 1 because of the 'universe' joint
```
This comment is outdated, maybe say 'number of generalized coordinates'.

buschbapti replied:
`get_number_of_joint_frames`? It is becoming a bit verbose

domire8 replied:
no no i was talking about the inline comment on line 223 not about function naming

buschbapti replied:
right good catch

domire8 reviewed (approved):
Looks good to me, nice!
On another note, I think you can create a model with the empty constructor `model = Model()`. Therefore, one can already call all the public functions even though the model is not initialized yet right? Is that a desired behaviour?

buschbapti commented:
@domire8 that is a very good point. I think the empty constructor for model is needed only for copying. I am not even sure. We could make it private if needed. Or we could also check for initialization and throw an error in every functions but this is tedious.

buschbapti commented:
Ok I have just removed the urdf with tool and we are good to go
